### ISSUE TITLE
Implement flash decoding

### ISSUE THREAD

**Issue by pfultz2** [ROCm/AMDMIGraphX#4334]
Implement flash decoding as described here: https://pytorch.org/blog/flash-decoding/

We have attention operators grouped like this:

```
Q -> [B, M, k]
K -> [B, k, N]
V -> [B, N, D]

S = dot(Q, K)
P = softmax(S)
O = dot(P, V) # [B, M, D]
```

To do flash decoding we will need to add another batch dimension for each group we want to split, and then do:

```
Q -> [B, G, M, k] # G is a broadcasted dimension
K -> [B, G, k, N/G]
V -> [B, G, N/G, D]

# first kernel
S = dot(Q, K)
P = softmax(S, axis=-1)
L = LSE(S) # [B, G, M, 1]
O' = dot(P, V) # [B, G, M, D]

# second kernel
scale = softmax(L, axis=1) # [B, G, M, 1]
R = mul(O', broadcast(scale)) # [B, G, M, D]
O = sum(R, axis=1) # [B, 1, M, D]
```

**Comment by pfultz2** [comment]
Also, another reference we used when we were discussing the flash decoding: https://arxiv.org/pdf/2402.05099

**Comment by bdevorem** [comment]
rocMLIR experiments for good G sizes: [rocMLIR#1895](https://github.com/ROCm/rocMLIR/pull/1895)

**Comment by bdevorem** [comment]
corrected kernel 2:
```
# second kernel
m = max(L, axis=1) # [B, 1, M, 1],  global max for stability
alpha = exp(L - m) # [B, G, M, 1],  unnormalized weights
scale = alpha / sum(alpha, axis=1) # [B, G, M, 1],  normalized weights
R = mul(O', scale) # [B, G, M, D]
O = sum(R, axis=1) # [B, 1, M, D]
```
since each group's output O'[g] is already normalized by its own softmax
